pulp_python/app/tasks/publish.py

```diff
@@ -107,39 +107,21 @@ def write_simple_api(publication):
     )
     index_metadata.save()
 
-    def find_artifact():
-        _art = content_artifact.artifact
-        if not _art:
-            _art = models.RemoteArtifact.objects.filter(content_artifact=content_artifact).first()
-        return _art
-
-    for (name, canonical_name) in index_names:
-        project_dir = '{simple_dir}{name}/'.format(simple_dir=simple_dir, name=canonical_name)
+    releases = python_models.PythonPackageContent.objects.filter(pk__in=publication.repository_version.content).values("name", "filename", "contentartifact", "_artifacts__sha256").order_by("name")
+    release_content_artifacts = python_models.PythonPackageContent.objects.filter(pk__in=publication.repository_version.content).values_list("contentartifact", flat=True)
+    remote_artifacts = models.RemoteArtifact.objects.filter(content_artifact__in=release_content_artifacts).values_list("content_artifact", "sha256").iterator()
+    checksums = {ca: sha for ca, sha in remote_artifacts}  # This can grow to 4 million elements if fully PyPI synced
```

Comment:
The `remote_artifacts` query can be removed (along with `release_content_artifacts` although that query shouldn't be evaluated) if we add `sha256` to the `PythonPackageContent` model.  We have to get the remote artifacts incase the sync is `on_demand` because the download links require the `sha256` be present.  While the query has been optimize to be quick and take up little memory the dictionary `checksums` that I create from it is really large and can grow to be up to 4 million elements (number of releases available on PyPI). I need to do more tests to check how bad the memory usuage is, but I have a feeling this optimization is needed to prevent going OOM.

Reply:
Potentially we should do that anyways if filename is not a workable (global) uniqueness constraint.